Add reduced diagnostics for beam

n01r opened the pull request:
Calculate and print mean values and standard deviations for positions and momenta. 
Derive RMS emittances from those and also print them.

So far, this only prints out the calculated values into the standard output.
However, the values are consistent with the analysis from the test of the FODO cell.

This is the first part for closing #269.

**Edit 2023-04-18: To Do**
- [x] Add Python bindings
- [x] Add documentation for output file

n01r committed:
Print out reduced beam diagnostics

Calculate and print mean values and standard deviations for positions and momenta.
Derive RMS emittances from those and also print them.

n01r commented on src/particles/diagnostics/BeamRelevant.cpp:
```diff
@@ -0,0 +1,225 @@
+                        const ParticleReal p_uy_mean = p_uy*p_w;
+                        const ParticleReal p_ut_mean = p_ut*p_w;
+                        const ParticleReal p_gm_mean = std::sqrt(1.0_rt+p_us*inv_c2)*p_w;
+
```
alpha and beta (CS parameters) are still missing

cemitch99 commented on src/particles/diagnostics/BeamRelevant.cpp:
```diff
@@ -0,0 +1,225 @@
+                        const ParticleReal p_ux_mean = p_ux*p_w;
+                        const ParticleReal p_uy_mean = p_uy*p_w;
+                        const ParticleReal p_ut_mean = p_ut*p_w;
+                        const ParticleReal p_gm_mean = std::sqrt(1.0_rt+p_us*inv_c2)*p_w;
```
I'm not sure about this line.  What is p_gm intended to represent?  Internally, the momenta should be dimensionless, so I don't think 1 + p_us*inv_c2 can work for dimensional reasons.

n01r replied:
Thanks! I removed those :)

cemitch99 commented on src/particles/diagnostics/BeamRelevant.cpp:
```diff
@@ -0,0 +1,225 @@
+                        const ParticleReal p_uy = ptd.m_rdata[RealSoA::uy][i];
+                        const ParticleReal p_ut = ptd.m_rdata[RealSoA::pt][i];
+                        const ParticleReal p_us = p_ux*p_ux + p_uy*p_uy + p_ut*p_ut;
+                        const ParticleReal p_gm = std::sqrt(1.0_rt+p_us*inv_c2);
```
The same comment applies here regarding p_gm.

cemitch99 commented on src/particles/diagnostics/BeamRelevant.cpp:
```diff
@@ -0,0 +1,225 @@
+                        // prepare mean square for momenta
+                        const ParticleReal p_ux_ms = (p_ux-ux_mean)*(p_ux-ux_mean)*p_w;
+                        const ParticleReal p_uy_ms = (p_uy-uy_mean)*(p_uy-uy_mean)*p_w;
+                        const ParticleReal p_uz_ms = (p_ut-ut_mean)*(p_ut-ut_mean)*p_w;
```
Is there any advantage to switching notation from ut to uz here?  It's fine, but it may be confusing to those reading the code.

n01r replied:
Fixed now!

cemitch99 commented on src/particles/diagnostics/BeamRelevant.cpp:
```diff
@@ -0,0 +1,225 @@
+            ParticleReal emittance_x = std::sqrt(x_ms*ux_ms-xux*xux);
+            ParticleReal emittance_y = std::sqrt(y_ms*uy_ms-yuy*yuy);
+            ParticleReal emittance_z = std::sqrt(t_ms*ut_ms-tut*tut);
+
```
More useful to users will be the normalized emittance, so at some stage we should multiply emittance_x,y,z by the reference value of beta*gamma.  The emittance values will then differ from those currently used in the benchmark validation scripts, so we should keep this in mind.

n01r replied:
Thanks, @cemitch99! I am printing out beta*gamma now as well so that we can just compute this offline, if that's okay. :)
This pull request is going to be an intermediate step towards the output in openPMD but if anyone is going to use it during this stage, we have it available now as well. 👍

cemitch99 commented on src/particles/diagnostics/BeamRelevant.cpp:
```diff
@@ -0,0 +1,225 @@
+                               << charge << " "
+                               << "\n";
+
+        }
```
This output looks great.  In IMPACT-Z, we output these quantities, except that the reference value of "s" replaces "step" in the first column, since users are interested in the beam moments as a function of lattice location.

n01r replied:
Thanks for the suggestion! I just added "s" to the output and kept the steps, just so we have both available for testing or debugging. :)

cemitch99 reviewed (commented):
The math looks good to me.  I added a few minor comments.

n01r committed:
Add Twiss parameters to prints and fix CmakeLists

- add Courant-Snyder (Twiss) alpha and beta
- remove the gamma because we are using normalized momenta
- put BeamRelevant.cpp to the right CMakeLists.txt as a target

n01r committed:
Remove unused variables

n01r committed:
Refactor naming and output into files

- renamed BeamRelevant (like in WarpX) to ReducedBeamCharacteristics
- call the calculation in a second function that handles only the output
- replace function from printing to returning calculated values

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.H:
```diff
@@ -0,0 +1,61 @@
+#define IMPACTX_REDUCED_BEAM_CHARACTERISTICS_H
+
+#include "particles/ImpactXParticleContainer.H"
+#include "ImpactX.H"
```
I think this is an unused import.
```suggestion
```

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.H:
```diff
@@ -0,0 +1,61 @@
+{
+    /** Struct containing reduced beam characteristics data
+     */
+     struct ReducedBeamCharacteristics{
```
One could also use a `std::unordered_map<std::string, amrex::ParticleReal>` instead. That way, the Python bindings are a `dict` and one does not need to bind another type.

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.H:
```diff
@@ -0,0 +1,61 @@
+
+#include <cmath>
+
+using namespace amrex;
```
Do not use namespaces in `.H` files - this will "pollute" the include space of every downstream file that includes it.
```suggestion
```

Please only use this in `.cpp` files and even better, only in `{ ... }` sections of functions when needed at all.

ax3l replied:
The only thing we really use in `{ ... }` is `using namespace amrex::literals; // for _rt and _prt`. The rest, please always write `amrex::` for clarity.

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -0,0 +1,208 @@
+#include <AMReX_ParallelDescriptor.H>   // for ParallelDescriptor
+#include <AMReX_BLProfiler.H>           // for TinyProfiler
+
+using namespace amrex;
```
We don't need this even in C++ files, `amrex::` is short.
```suggestion
```

n01r committed:
Return type unordered map and add Python bindings

- Change return type of the function to unordered map
- add python bindings and add the function to a test

n01r commented:
The new commit adds the Python bindings, and I tested already that they work.
Documentation is still missing.

n01r commented on tests/python/test_impactx.py:
```diff
@@ -26,6 +26,11 @@ def test_impactx_fodo_file():
 
     sim.evolve()
+    # test calculation of reduced beam characteristics
+    reduced_beam_data = sim.reduced_beam_characteristics()
```
@ax3l, do we want a separate test or is it fine to include this here?

ax3l replied:
Here is good, please add a validation for the final beam here as done in these lines:
https://github.com/ECP-WarpX/impactx/blob/c29f1ec3f6261963c45b1063afac8d496ebd9e7f/examples/fodo/analysis_fodo.py#L88-L104

n01r replied:
Thanks! Final beam validation was added

ax3l commented on src/python/ImpactX.cpp:
```diff
@@ -6,6 +6,7 @@
 #include "pyImpactX.H"
 
 #include <ImpactX.H>
+#include "particles/diagnostics/ReducedBeamCharacteristics.H"
```
See code around your patch :)
```suggestion
#include <particles/diagnostics/ReducedBeamCharacteristics.H>
```

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.H:
```diff
@@ -0,0 +1,30 @@
+#include <AMReX_REAL.H>
+#include <AMReX_GpuComplex.H>
+
+#include <cmath>
```
```suggestion
#include <cmath>
#include <string>
#include <unordered_map>

```

n01r committed:
Add beam parameter validation

- add reduced beam characteristic validation
- remove unused include

n01r committed:
Rename all momenta from udim to pdim

n01r committed:
Rebase and fix naming of momenta

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

n01r committed:
Add documentation for reduced_beam_characteristics

n01r committed:
Remove unused include to avoid build error in CI

Removed `#include "particles/NamedComponentParticleContainer.H"`

n01r committed:
Remove comment from old code

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -0,0 +1,224 @@
+            auto r = amrex::ParticleReduce<amrex::ReduceData<amrex::ParticleReal,amrex::ParticleReal,
+            amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal>>(
+                    pc,
+                    [=] AMREX_GPU_DEVICE(const PTDType& ptd, const int i) noexcept -> amrex::GpuTuple
+                            <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal,
+                            amrex::ParticleReal,amrex::ParticleReal>
+                    {
+                        // access AoS particle position data
+                        const amrex::ParticleReal p_pos0 = ptd.m_aos[i].pos(0);
+                        const amrex::ParticleReal p_pos1 = ptd.m_aos[i].pos(1);
+                        const amrex::ParticleReal p_pos2 = ptd.m_aos[i].pos(2);
```
Unless I am mistaken, you cannot access `ptd.m_aos` as a whole here.
Use the pattern we use in WarpX for now please.
https://github.com/ECP-WarpX/WarpX/blob/872bae2026a6767f2023f8ba600c87d247eed38e/Source/Diagnostics/ReducedDiags/BeamRelevant.cpp#L198-L209

n01r replied:
> Unless I am mistaken, you cannot access ptd.m_aos as a whole here.   

Sorry, could you explain? 🤔  It works on MacOS and on Perlmutter with all tests that I did. 
I was trying the pattern you mentioned before but the one I am using now is the one that @atmyers suggested.

ax3l replied:
As discussed on Slack: could be a compiler bug for the NVCC version in CI. Please try again w/ the syntax in the link above.

n01r committed:
Add missing includes suggested via PR review

n01r committed:
AoS - SuperParticleType not ParticleTileDataType

ax3l committed:
CUDA: Drop <11.3

Drop support for older CUDA releases. The CUDA Toolkit 11.0
added C++17 support but had a few severe issues with it, which
are better in 11.3+. These releases are now widely available,
so we can drop older releases.

ax3l commented on docs/source/dataanalysis/dataanalysis.rst:
```diff
@@ -3,8 +3,38 @@
+
 .. note::
 
    TODO :-)
```
```suggestion
   openPMD beam monitor output: TODO :-)
```

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -113,6 +113,11 @@ General
 
+   .. py:method:: reduced_beam_characteristics()
+
+      Compute reduced beam characteristics like the position and momentum moments of the particle distribution, as well as emittance and Twiss parameters.
```
```suggestion
      Compute reduced beam characteristics like the position and momentum moments of the particle distribution, as well as emittance and Twiss parameters.

      :return: beam properties with string keywords
      :rtype: dict
```

ax3l committed:
Includes & Formatting

ax3l committed:
Doc: Return type

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/particles/diagnostics/DiagnosticOutput.cpp:
```diff
@@ -42,9 +43,35 @@ namespace impactx::diagnostics
+            std::unordered_map<std::string, amrex::ParticleReal> const rbc = diagnostics::reduced_beam_characteristics(
+                    pc);
+
+            file_handler << step << " " << rbc.at("s") << " " << rbc.at("beta_gamma") << " "
```
```suggestion
            file_handler << step << " " << rbc.at("s") << " " << rbc.at("ref_beta_gamma") << " "
```

ax3l commented on src/particles/diagnostics/DiagnosticOutput.cpp:
```diff
@@ -42,9 +43,35 @@ namespace impactx::diagnostics
             } else if (otype == OutputType::PrintRefParticle) {
                 file_handler << "step s x y z t px py pz pt\n";
+            } else if (otype == OutputType::PrintReducedBeamCharacteristics) {
+                file_handler << "step" << " " << "s_ref_part" << " " << "beta_gamma_ref_part" << " "
```
```suggestion
                file_handler << "step" << " " << "s" << " " << "ref_beta_gamma" << " "
```

ax3l commented on docs/source/dataanalysis/dataanalysis.rst:
```diff
@@ -3,8 +3,38 @@
+
+* ``step``
+    Iteration within the simulation
+* ``s_ref_part``, ``beta_gamma_ref_part``
```
```suggestion
* ``s``, ``ref_beta_gamma``
```

ax3l committed:
Rename Columns

ax3l committed:
Charge in Coulomb

ax3l committed:
[Hack] Remove `noexcept` on lambdas

ax3l commented:
Looking at the error message, this is CUDA 11.0 but looks similar to a bug in CUDA 11.4-11.8
https://github.com/pybind/pybind11/issues/4193#issuecomment-1270491939
```
... <char,     ::std::char_traits<char> , ::std::allocator<char> > , float> > >  (*)(const  ::impactx::ImpactXParticleContainer &),(& :: impactx::diagnostics::reduced_beam_characteristics[abi:cxx11]), 1> >,const  ::amrex::ConstParticleTileData< ::amrex::Particle<(int)0, (int)0> , (int)5, (int)0> >>> 
```

Update: ah no, actually the casts here are ok-ish, I think.
Maybe a problem with `& :: impactx::diagnostics::reduced_beam_characteristics[abi:cxx11]` or something else from the code transformation in that CUDA version. Testing if bumping to 11.3+ via #361 helps.

ax3l committed:
Revert "[Hack] Remove `noexcept` on lambdas"

This reverts commit b07d51c9466455db38099c27321428edadefee40.

ax3l committed:
Merge remote-tracking branch 'mainline/development' into topic-AddReducedBeamDiagnostics

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -0,0 +1,251 @@
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "ImpactX.H"
+#include "particles/ImpactXParticleContainer.H"
+#include "particles/diagnostics/ReducedBeamCharacteristics.H"
```
Own import always first + `ImpactX.H` not used:
```suggestion
#include "ReducedBeamCharacteristics.H"

#include "particles/ImpactXParticleContainer.H"
#include "particles/ReferenceParticle.H"
```

ax3l committed:
Fix includes

ax3l reviewed (approved):
Thank you, this is great! :sparkles:
